### PR TITLE
Remove MappedNode & generalise render logic

### PR DESCRIPTION
**Pull request by tomwilkie**
Towards #228

### CODE REVIEW

**Comment by peterbourgon** [comment]
Could you write a few sentences for context? At a minimum, the problem that motivates this PR, and a a high-level description of the implementation?

**Comment by tomwilkie** [comment]
This is a refactoring on the path to synthesising the containers topology from endpoints topology via the process topology.  To do this we need the renderers to be able to recurse - effectively what I'm aiming for is:

```
render.Reduce{
	render.LeafMap{Selector: report.SelectContainer, Mapper: report.MapContainerIdentity, Pseudo: report.InternetOnlyPseudoNode},
	render.Map{
		Mapper: report.MapProcess2Container,
		Renderer: render.Reduce{
			render.LeafMap{Selector: report.SelectProcess, Mapper: report.MapProcessIdentity, Pseudo: report.InternetOnlyPseudoNode},
			render.Map{
				Mapper: report.MapEndpoint2Process,
				Renderer: render.LeafMap{Selector: report.SelectEndpoint, Mapper: report.MapEndpointIdentity, Pseudo: report.InternetOnlyPseudoNode},
			}
		}
	}
}
```

This isn't all the way there yet, just a step in that direction (getting the Renderer interface right I think).

I removed MappedNode as its redundant, with the introduction of DetailedNode.Merge.  It has the nice side effect of propagating Origin nodes correctly for nodes without any adjacency info.

File: render/render.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package render
+
+import (
+	"github.com/weaveworks/scope/report"
+)
+
+// Renderer is something that can render a report to a set of RenderableNodes
+type Renderer interface {
+	Render(report.Report) report.RenderableNodes
+	EdgeMetadata(rpt report.Report, localID, remoteID string) report.AggregateMetadata
```
Comment (peterbourgon) [inline render/render.go]: This should be called AggregateMetadata, no?

File: render/render_test.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+type mockRenderer struct {
+	nodes report.RenderableNodes
+	amd   report.AggregateMetadata
+}
+
+func (m mockRenderer) Render(rpt report.Report) report.RenderableNodes { return m.nodes }
+func (m mockRenderer) EdgeMetadata(rpt report.Report, localID, remoteID string) report.AggregateMetadata {
+	return m.amd
+}
```
Comment (peterbourgon) [inline render/render_test.go]:
```go
type mockRenderer struct {
    report.RenderableNodes
    report.AggregateMetadata
}

func (m mockRenderer) Render(report.Report) report.RenderableNodes { 
    return m.RenderableNodes 
}

func (m mockRenderer) XxxMetadata(report.Report, string, string) report.AggregateMetadata { 
    return m.AggregateMetadata 
}
```

File: report/mapping.go [hunk 3]
```diff
@@ -25,13 +39,13 @@ type MappedNode struct {
 //
 // If the final output parameter is false, the node shall be omitted from the
 // rendered topology.
-type MapFunc func(string, NodeMetadata) (MappedNode, bool)
+type MapFunc func(NodeMetadata) (RenderableNode, bool)
```
Comment (peterbourgon) [inline report/mapping.go]: Nice.

**Comment by peterbourgon** [comment]
This feels really refreshing. Whenever we can eliminate a needless intermediate form, wonderful. Thanks! LGTM.

File: render/render_test.go (hunk 2 above)
Comment (tomwilkie) [inline render/render_test.go]: Can't have field and method called AggregateMetadata.  But otherwise it works.

**Pull request merged**